### ISSUE TITLE
Renaming ports

### ISSUE THREAD

**Issue by szszoke** [geoffreybennett/alsa-scarlett-gui#25]
Renaming ports would be really helpful when a 18i20 is used as a standalone mixer/patchbay.

What would it take to implement this?

As far as I remember the ports can be renamed in Focusrite Control but I'm not if that is implemented with the current ALSA driver.

**Comment by geoffreybennett** [comment]
I agree, renaming ports would be really useful, and you are right, the ports can be renamed in FC but it's not implemented in the ALSA driver. I think the driver needs a way to access the rest of the configuration space on the device as read/write bytes (no interpretation), so the remaining software-side features can be implemented. E.g. stereo linking, panning, naming channels, etc.

**Comment by ironhouzi** [comment]
Hi @geoffreybennett . Thanks for taking your time to develop these tools that allow us to use proper audio interfaces to their full extent in Linux! I have made a donation as a small token of appreciation and was hoping to gain insight into how stereo linking and port renaming could be developed.

I can fully appreciate implementing the feature where it belongs and I see that it is claimed that the ALSA driver is the correct place to implement such features. However, it seems that getting such feature into ALSA [seems tricky](https://github.com/Focusrite-Scarlett-on-Linux/Information/discussions/6).

From a naive standpoint, I would assume that these features could be solved in user space. Having a simple table lookup from a user defined set of keys to the ALSA specified port names. The stereo linking I was hoping to have the user be able to link dials and implement logic that mirrors the values.

If such a solution is less than ideal, would it at least serve as a 100% usable workaround until the ALSA devs find the proper solution?

**Comment by geoffreybennett** [comment]
Hi @ironhouzi,

Thank you for your support and encouragement! It's really appreciated and helpful to me!

I'm not sure that getting these features into ALSA would be tricky; the infrastructure needed all appears to be already there in the ALSA core, and only changes to my scarlett2 driver would be needed (as far as I can tell!).

The problem I've been struggling with for the past years is where do we store this info so that it is not lost when the interface is powered off? I really do like how you can turn the interface off, move it to another computer, and its entire configuration is still retained within the device and retrieved by the driver when you plug it in.

There is a "software configuration space" used by Focusrite Control to store most of this stuff for most models. But "most" is not all. The Linux driver supports more hardware features across more models than FC does, and the way the FC uses that space is inadequate to describe all the ways that hardware can be configured.

Possibly I could repurpose the software configuration space, but I would want some advice from the Focusrite firmware developers on this, and reassurance that any future firmware updates wouldn't break this in the future. But I understand that they're rather busy right now.

Your naive standpoint is valid; these features could definitely be implemented in user space and that was what I was leaning towards. At the moment, alsa-scarlett-gui is only an interface to the kernel controls. But there's no reason that it could not maintain additional information, stored on disk somewhere. It would need to handle the two getting out of sync, but I don't think that's an unsolvable problem.

The problem with implementing it entirely in user-space is then that applications like this: https://github.com/geoffreybennett/alsa-scarlett-gui/discussions/119 https://github.com/geoffreybennett/alsa-scarlett-gui/discussions/93 https://github.com/geoffreybennett/alsa-scarlett-gui/issues/92 would not benefit.

Upon further reflection, I'm thinking that a hybrid approach is necessary:
1) Implement stereo-pairing, pan, mute, solo in the driver.
2) If possible (future problem), save the state of the additional controls within the interface.
3) If the additional controls' state can't be read, then on driver startup default to mono, etc. for backwards compatibility.
4) If the additional controls' state can't be read, then alsactl or alsa-scarlett-gui can put back the enhanced controls' state from its own saved state on disk.

Regards,
Geoffrey.

**Comment by ironhouzi** [comment]
Hi @geoffreybennett . Thank you for the detailed response.

I think you are absolutely correct in wanting portable configuration and I fully agree that it is the most elegant solution.

If I understand correctly, this ideal approach hinges on a requirement that relies on undefined behavior. I deem it undefined, since Focusrite has made no obligation to uphold any standards in terms of how device storage is to be managed. Even if you get insight into a safe way to write data to the device from some of their devs, there is no guarantee that these current practices will be upheld in the future. Their internal procedures can change if current management, engineers change jobs, or perhaps a change in share holders force a change in their engineering practices.

I think storing to disk the custom port names, and stereo channels is much simpler in implementation and this configuration can be synchronized between devices using, dropbox, syncthing, git, dotfile-manager, etc. and it would not rely on undefined behavior. I think the amount of users that would benefit from stereo channels and custom port names heavily outweighs their expectancy that these settings remain portable.

Also, I would challenge the notion that a common scenario is to regularly move the audio interface between multiple computers. If you want to design features for your own use case, that is completely valid and I would fully support that decision. However, if you are hoping to benefit the majority of your users, it might be a good idea to get some data points as to how many users move their interfaces around, since it would directly point to the design choices of these features.

I want to make it completely clear that I am in no way expecting you to do anything other than what you deem sensible. Your generosity in time and effort earns you the right to manage this software as you seem fit. I will support your decision regardless and merely hope you will earnestly take my feedback into consideration.

Thank you so much for caring about these matters!

**Comment by szszoke** [comment]
I would also agree with @ironhouzi that even if port names/stereo pairs are local only, it would still bring value to those (of us) who doesn't move sound cards between devices.

**Comment by geoffreybennett** [comment]
Hi @ironhouzi, @szszoke,

Oh my, only 4 years(!), but port renaming is now available in the 1.0beta branch via the Configuration window (Ctrl-G). I went with the user-space approach you suggested; state is stored locally in ~/.config/alsa-scarlett-gui/ per device. It doesn't travel with the device if you move it to another PC, but the feature works now and doesn't extend the use of the flash in ways that were completely unintended by the FR firmware developers.

Also implemented: default names, stereo linking across all I/O types, and show/hide for unused ports (all stored the same way), and an oodle of other useful things.

<img width="945" height="551" alt="Image" src="https://github.com/user-attachments/assets/47adef30-6f35-435f-9ef1-88d472e12582" />

The code is pretty close to what it should be for the 1.0 release, but I still need to test on every supported device (28(!) of them across 7 product series) and I keep finding more little things to fix. Also, the docs still need a heap of work. Feedback welcome!

Regards,
Geoffrey.

**Comment by szszoke** [comment]
Hey,

Thanks! I will try to take a look at this over the weekend.

**Comment by szszoke** [comment]
I gave it an initial try with my 18i20 MK3 and the port names seem to be recalled just fine! I will try to test it a bit more  and clean up some unused ports, now that they can be hidden.
